internal/client/db/pinned_blocks_cache.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package db
+
+import (
+	"log"
+	"math"
+
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	lru "github.com/hashicorp/golang-lru/v2"
+)
+
+// Entry for pinned blocks cache.
+type pinnedBlocksCacheEntry struct {
+	// How many times this item has been pinned
+	refCount uint32
+
+	// Cached justifications for this block
+	Justifications runtime.Justifications
+
+	// Cached body for this block
+	Body *[]runtime.Extrinsic
+}
+
+func (pbce *pinnedBlocksCacheEntry) DecreaseRef() {
+	if pbce.refCount > 0 {
+		pbce.refCount--
+	}
+}
+
+func (pbce *pinnedBlocksCacheEntry) IncreaseRef() {
+	if pbce.refCount < math.MaxUint32 {
+		pbce.refCount++
+	}
+}
+
+func (pbce *pinnedBlocksCacheEntry) HasNoReferences() bool {
+	return pbce.refCount == 0
+}
+
+// Reference counted cache for pinned block bodies and justifications.
+type pinnedBlocksCache[H comparable] struct {
+	cache *lru.Cache[H, *pinnedBlocksCacheEntry]
+}
+
+func newPinnedBlocksCache[H comparable]() pinnedBlocksCache[H] {
+	cache, err := lru.NewWithEvict[H, *pinnedBlocksCacheEntry](1024, func(key H, value *pinnedBlocksCacheEntry) {
+		// If reference count was larger than 0 on removal,
+		// the item was removed due to capacity limitations.
+		// Since the cache should be large enough for pinned items,
+		// we want to know about these evictions.
+		if value.refCount > 0 {
+			log.Printf("TRACE: Pinned block cache limit reached. Evicting value. hash = %v\n", key)
```

Comment:
Should we not be using our existing logger to match up with the rest of our code? So in this case use `logger.Tracef(...)`

Reply:
I think I may do the same thing as the standalone grandpa package and use a local interface for logging.  I'd rather not import our existing logger directly.

Reply:
I am not sure why this is a better solution than using the existing logger, could you elaborate? Or if we are going to revisit later this could be another item to add to the issue tracking work to do before integrating/merging to development

Reply:
I think representing the package logger as an interface is better than actually referencing a concrete logger type. At this stage of my development though, I would still print out TRACE logs to stdout while I'm in active development.  But this will definitely be accounted for before merging this feature branch.

Reply:
gotcha I see, this makes sense to me